### PR TITLE
Chunks

### PR DESCRIPTION
**Pull request by orianac**
Three changes:
* Building out some checks in the chunks schema. 
* Adding a check that the input is actually a da, since if it's a dataset the chunk routines will fail for the wrong reason.
* Remove the return of the da at the end since I think we just want it to continue without returning anything.

### CODE REVIEW

File: xarray_schema/core.py [hunk 1]
```diff
@@ -106,9 +107,13 @@ def validate(self, da: xr.DataArray) -> xr.DataArray:
             dim_chunks = dict(zip(da.dims, da.chunks))
             for key, ec in self.chunks.items():
                 if isinstance(ec, int):
-                    for ac in dim_chunks[key][:-1]:
-                        if ac != ec:
-                            raise SchemaError(f'{key} chunks did not match: {ac} != {ec}')
+                    # handles case of expected chunksize is shorthand of -1 which translates to the full length of dimension
+                    if ec==-1:
```
Comment (jhamman) [inline xarray_schema/core.py:111]: any negative value
```suggestion
                    if ec < 0:
```

**Review by jhamman — CHANGES_REQUESTED** [review]
Thanks @orianac - looks like a good start. This also will need a test. Actually at least 3 ;)

1. test your expected behavior with chunk=-1
2. test that it raises an exception when chunk=-1
3. test that a positive integer raises an error with uneven chunking (I can help if needed with this one)

File: xarray_schema/core.py [hunk 2]
```diff
@@ -106,9 +107,13 @@ def validate(self, da: xr.DataArray) -> xr.DataArray:
...
+                        ec = len(da[key])
```
[the remark this answers is not in the document]
Comment (orianac) [reply xarray_schema/core.py:113]: implemented!

File: xarray_schema/core.py [hunk 3]
```diff
@@ -106,9 +107,13 @@ def validate(self, da: xr.DataArray) -> xr.DataArray:
...
+                        # grab the first entry in da's tuple of chunks to be representative (as it should be assuming they're regular)
+                    ac = dim_chunks[key][0]
```
[the remark this answers is not in the document]
Comment (orianac) [reply xarray_schema/core.py:115]: i'll check by pulling in the dask utility for that `_check_regular_chunks`

File: tests/test_core.py [hunk 4]
```diff
@@ -125,3 +145,9 @@ def test_dataset_example():
         }
     )
     ds_schema.validate(ds)
+
+def test_validate():
```
[the remark this answers is not in the document]
Comment (jhamman) [reply tests/test_core.py:149]: I see what you're going for but I think we need a bit more API to make it happen. What if you instead wrote

```python
schema = DataArraySchema(chunks=False)
```

Whereas `chunks=None` currently indicates that the schema doesn't specify any kind of chunking, `chunks=False` would specify the DataArray be "unchunked".  Conversely, `chunks=True` would require the array to be chunked but otherwise not care about the chunk layout.

File: xarray_schema/core.py [hunk 5]
```diff
@@ -72,6 +72,11 @@ def validate(self, da: xr.DataArray) -> xr.DataArray:
         ------
         SchemaError
         '''
+        if not isinstance(da, xr.DataArray):
+            raise ValueError('Input must be a xarray.DataArray')
+        
+        if da.chunks:
+            assert _check_regular_chunks(da.chunks), 'Good gracious no! Chunks are not regular!'
```
Comment (jhamman) [inline xarray_schema/core.py:79]: irregular chunks are "fine" in some cases. So I don't think we want to raise an error here. I'd rather catch this down in the `chunks` section if regular chunks are specified.
```suggestion
```

File: xarray_schema/core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jhamman) [reply xarray_schema/core.py:119]: Here I think we need to check all the chunk values and allow the last chunk to be `<= ec`. Something like this

```python
if not (all([ac == ec for ac in dim_chunks[key][:-1]) and dim_chunks[key][-1] <= ec):
    raise SchemaError(f'{key} chunks did not match: {ac} != {ec}')

File: xarray_schema/core.py [hunk 6]
```diff
@@ -124,7 +134,7 @@ def validate(self, da: xr.DataArray) -> xr.DataArray:
         if self.checks:
             for check in self.checks:
                 da = check(da)
-
+
```
Comment (jhamman) [inline xarray_schema/core.py:137]:
```
pre-commit install
```
:)

**Review by jhamman — CHANGES_REQUESTED** [review]
Getting oh so close

File: xarray_schema/core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (orianac) [reply xarray_schema/core.py:119]: I do that now by using dask's `_check_regular_chunks` (https://github.com/dask/dask/blob/a135dd96e1ec6a56f4c80a1c3ad29d50caabe848/dask/array/core.py#L3413) in L79 before any of the other checking gets started. It seems useful to check it. If we don't want to force checking for it we could add a top-level flag like `allow_irregular_chunks`?

File: tests/test_core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (orianac) [reply tests/test_core.py:149]: This is interesting. I am noodling. 🍝

[the remark this answers is not in the document]
Comment (orianac) [reply tests/test_core.py:149]: Now we allow for chunks to either be a `bool`, a `dict`, or `None`. The case of `chunks=False` gets checked in `test_validate`, and the rest get checked in `test_dataarray_validate_chunks`.

File: xarray_schema/core.py (hunk 6 above)
Comment (orianac) [reply xarray_schema/core.py]: 🧹

**Comment by orianac** [comment]
@jhamman Take a look- added logic to cover chunks being `True`, `False` or a dict, and added a few more tests to cover some corner cases.

**Comment by jhamman** [comment]
Thanks @orianac. I'm going to merge now!
